### DISCUSSION TITLE
Which functions need to be used to be still undetectable with uc=True

### DISCUSSION THREAD

**Discussion by MacMarde** [seleniumbase/SeleniumBase#2229] [Q&A] [answered]
@mdmintz 
As we know it is not possible to enter the URL automatically into the browser. If you do so, you will no longer pass CF captcha here: [nowsecure.nl](nowsecure.nl)
Therefore you need to use the `driver.get("https://www.nowsecure.nl")` function.

See [here](https://github.com/seleniumbase/SeleniumBase/discussions/2043#discussioncomment-6889840)

So the question is, are there more functions which should not be performed manually, for example sendkeys, click an element etc. ?

I want to do some tests to see if the whole SeleniumBase uc-driver is working for me before automating everything.
The idea is to start a browser with uc=True in SB and then just perform as much actions manually as possible without risking being detected.

**Comment by mdmintz** [comment, marked as answer]
Hello, In UC Mode (`--uc` / `uc=True`) SeleniumBase's `driver.get(url)` function has been modified to disconnect chromedriver from Chrome before navigating to the URL. This disconnect is necessary to prevent detection. There are other special methods, such as `driver.uc_click(selector)` and `driver.uc_open_with_reconnect(url, reconnect_time)`, etc.

See https://www.youtube.com/watch?v=5dMFI3e85ig for details.

**Comment by MacMarde** [comment]
> Hello, In UC Mode (`--uc` / `uc=True`) SeleniumBase's `driver.get(url)` function has been modified to disconnect chromedriver from Chrome before navigating to the URL. This disconnect is necessary to prevent detection. There are other special methods, such as `driver.uc_click(selector)` and `driver.uc_open_with_reconnect(url, reconnect_time)`, etc.
> 
> See https://www.youtube.com/watch?v=5dMFI3e85ig for details.

Thanks so much. I am on it.

**Comment by MacMarde** [comment]
@mdmintz 
Many clicks in my project on website elements opens a new website in a new browser tab.
How to make sure the new loading website in the new browser tab is also undetected?
As far as I understood you in your youtube video, `driver.uc_click(selector)` only disconnects for a short time. See [here](https://www.youtube.com/watch?v=5dMFI3e85ig&list=FLki3wsqfwOuzq_63rBNFjBQ&index=1&t=735s).
So how to handle it and make sure I am undetectable in this situations? Can I change the disconnect time for `driver.uc_click(selector)`?

**Reply by mdmintz** [reply]
You should avoid clicking to remain undetected if you can. Use the special methods, and set a high `reconnect_time` for `driver.uc_open_with_reconnect()` if you need to:

```python
driver.uc_open(url)
driver.uc_open_with_tab(url)
driver.uc_open_with_reconnect(url, reconnect_time=None)
```

`driver.uc_click(selector)` is configured to where it needs to be to work. Although expect improvements in the next version of SeleniumBase in several areas.

**Reply by MacMarde** [reply]
@mdmintz 
Many websites load new content or new URLs after clicking a button.
The problem is, how do I know when clicks and new URLs lead to anti-bot detection?
I do not know it and therefore I try to secure every click and every new website loading.
Do you have any hint for this?

How long is the current reconnect_time for `driver.uc_click(selector)`? Could it be used for my problem?

I would love to see a solution in the next version of SeleniumBase.

**Reply by mdmintz** [reply]
Lots of updates in `4.21.0`, now available: https://github.com/seleniumbase/SeleniumBase/releases/tag/v4.21.0

Try with the UC Mode examples, which include CF page loads, and clicking CAPTCHAs successfully:
* [SeleniumBase/examples/verify_undetected.py](https://github.com/seleniumbase/SeleniumBase/blob/master/examples/verify_undetected.py)
* [SeleniumBase/examples/raw_turnstile.py](https://github.com/seleniumbase/SeleniumBase/blob/master/examples/raw_turnstile.py)
* [SeleniumBase/examples/raw_form_turnstile.py](https://github.com/seleniumbase/SeleniumBase/blob/master/examples/raw_form_turnstile.py)

See each of those for advanced methods. If you have to perform actions after you get onto a website (while still remaining undetected) try the `driver.reconnect(seconds)` option.

**Reply by MacMarde** [reply]
How to use `driver.reconnect(seconds)`? It is not getting clear to me in the examples.

I mean what does this function do? 
Will it disconnect directly for 10 seconds after calling lets say `driver.reconnect(10)`.
And will I therefore be able to execute the next lines of code while being undetected?

So for example lets say I want to click a button that will open a new tab with a new URL.

driver.reconnect(10)
driver.click('someButtonID')

But how can I click a button when Selenium is disconnected?
I do not understand how it works and I do not understand how to use it correctly.
Please let me know it.

**Reply by MacMarde** [reply]
@mdmintz 
Have you seen my answer above?
I have tried the `driver.reconnect(10)` function, but it just does nothing for 10 seconds.
I am sorry I still have no idea how this could help.

**Reply by mdmintz** [reply]
That's how `driver.reconnect()` works. During that disconnected period, Chrome has no idea that it is (or ever was) controlled by chromedriver. Therefore, when websites you visit "ask" Chrome if it's been controlled by chromedriver, Chrome has no knowledge of it, and therefore the website thinks your browser is being controlled by a human.

As a side effect of this expected behavior, the link between chromedriver and Chrome is temporarily severed, and therefore chromedriver can't issue any commands to Chrome. This also means you have to guess how long to remain disconnected for, because there's no way to know for sure when a website is no longer asking Chrome if it's connected to chromedriver.

**Reply by MacMarde** [reply]
@mdmintz 
Ok thanks for the clarification. I think I understand it.
Sometimes after clicking a button, Selenium is waiting and pausing my python script until the new content has been loaded (probably not good for using this function).
But in cases where Selenium just clicks a button and immediately continues to process the python script it might help (when I am fast enough to call `driver.reconnect()` instantly).

**Reply by mdmintz** [reply]
You can set the page-load-strategy to "none" (`pls="none"`) to prevent some extra waiting. That could also lead to unintended consequences (flaky tests, etc.)

**Reply by MacMarde** [reply]
Ah yes you are right. Thanks for reminding me. I did not consider this.

**Reply by MacMarde** [reply]
@mdmintz 
I am thinking about a more dynamically way to solve this issue.
Because it is always just a matter of time when the reconnect time you set is not long enough when loading a website with `driver.uc_open_with_reconnect("https://www.ebay.com", 6)`
There could be many reasons for this. For example internet connection or proxy is slow, windows updates need high pc ressources, etc.

So I thought about something like this:
```
driver.uc_open_with_reconnect("https://www.ebay.com", 6)
while browser.execute_script("return document.readyState") != "complete":
        browser.reconnect(5)
```
Problem is that browser.reconnect(5) takes about 8 seconds in my test not 5 seconds. This means the rest of the function to disconnet and reconnect takes about 3 seconds plus 5 seconds timeout. The question is how long is the webdriver really connected in this example?
If it would be possible to make the actual time the webdriver is connected much shorter, you could use it to determine when the website has been completely loaded and the probability of getting detected this way would be less if the time being connected is for example within 100 ms.

What do you think? Can we measure the time actually connected? Can we make it faster?

**Reply by MacMarde** [reply]
It seems like it is not working as expected.
`browser.execute_script("return document.readyState")` sometimes returns "complete" when website is still loading and sometimes returns not instantly, instead it waits until website has been loaded completely and then also returns "complete".

I am just not happy with the UC solution. It is so easy for anti-bot measures to get around this solution. It would be great to have a reliable solution again like it was with the original UC from @ultrafunkamsterdam

**Comment by MacMarde** [comment]
@mdmintz 
Maybe it would be a great feature to have a function for connecting/disconnecting chromedriver manually whenever needed.
What do you think?

**Reply by mdmintz** [reply]
`driver.reconnect(seconds)` in UC Mode. Upgrade to the latest version.

**Comment by MacMarde** [comment]
@mdmintz 
What is with webelement.click() ?
Sometimes it is hard or not possible to get the element with a simple selector.

**Reply by mdmintz** [reply]
Use the built-in SeleniumBase methods for interacting with elements.

**Reply by MacMarde** [reply]
And it will still be undetectable?

**Reply by mdmintz** [reply]
Upgrade to `4.21.0` (or newer if available) and try the new UC Mode examples, which click CAPTCHAs successfully:
* [SeleniumBase/examples/raw_turnstile.py](https://github.com/seleniumbase/SeleniumBase/blob/master/examples/raw_turnstile.py)
* [SeleniumBase/examples/raw_form_turnstile.py](https://github.com/seleniumbase/SeleniumBase/blob/master/examples/raw_form_turnstile.py)

**Reply by MacMarde** [reply]
@mdmintz 
How can I click an element with uc_click() like this when there is more than 1 element available?:
`browser.find_elements(By.CLASS_NAME, 'table')[1].click()`

**Reply by mdmintz** [reply]
Use a better selector to uniquely identify the element you're trying to click.

Eg. https://seleniumbase.io/demo_page

```python
driver.uc_click('.checkBoxClassB:nth-of-type(3)')
```

(To click the 3rd checkbox from the group of checkboxes that all have class `checkBoxClassB`)

**Reply by MacMarde** [reply]
@mdmintz 
Could you please add an option for `driver.uc_click()` to customize the reconnect time within the function?
For example something ilke `driver.uc_click('.checkBoxClassB:nth-of-type(3)', reconnect = 10)`

I think it would be very usefull, because there are many website elements everywhere that trigger the load of new content or even new websites when clicked. This way you can always be safe and define a timeout that is long enough until all new content has been loaded.

What do you think?

**Reply by mdmintz** [reply]
That feature already exists and is already more than 5 days old.

https://github.com/seleniumbase/SeleniumBase/blob/f9a31a39c5544cbb5d4d290dcec9634fea14cddb/seleniumbase/core/browser_launcher.py#L438

Please make sure you are using the latest version of SeleniumBase before opening tickets.

**Reply by MacMarde** [reply]
@mdmintz 
What selectors are supported with SB? 
I can not click an element with its class name.

`browser.uc_click('center', by="class_name", reconnect_time = random.uniform(4, 6))    `

"class" and "class name" also not working.

**Reply by MacMarde** [reply]
@mdmintz 
Nevermind.
Something was weird.
It is working with "class name" instead of "class_name"

**Reply by MacMarde** [reply]
@mdmintz 

I have tried to select an element like in your example.
I need to select a span element that matches 4 class names. In this example I want to select the third element:

`driver.uc_click('span.class1.class2.class3.class4:nth-of-type(3)', by="css selector", reconnect_time = random.uniform(1, 2)) `

no matter what I am trying it is not working. Also not with xpath. It always gives me an error. element not found.

When using Selenium methods I find every element available on the website. In this case over 100 elements:

`browser.find_elements(By.XPATH, ('//span[@class = "class1 class2 class3 class4"]'))`

same true when using css selector.

I can select any element I want with Selenium function. Here for example element 25.

`browser.find_elements(By.XPATH, ('//span[@class = "class1 class2 class3 class4"]'))[25]`

But this does not help, because this way I think I can not parse it to uc_click()

Any idea how to do it? I tried for hours. No way to tell uc_click() which element to use.

**Reply by mdmintz** [reply]
Use a better CSS Selector. Or use XPath. Also your `random.uniform(1, 2)` will be less than 2, which might not be enough.

**Comment by MacMarde** [comment]
Could you please implement a way to use uc_click with elements like this:

webelement.uc_click()

It would be so much better and easier.

**Reply by mdmintz** [reply]
@MacMarde Have you tried that in UC Mode?

**Reply by iamumairayub** [reply]
@MacMarde AFAIK it already works 

`driver.find_element(by=By.CSS_SELECTOR, value='#some-id').uc_click(reconnect_time=5)`

**Reply by mdmintz** [reply]
Yes, `element.uc_click()` already exists:

https://github.com/seleniumbase/SeleniumBase/blob/7d9d5c5203e3d93bdfc50f9d113c88791e9ba26c/seleniumbase/undetected/webelement.py#L5

**Reply by MacMarde** [reply]
@iamumairayub 
Thanks so much. I will try it.
